### PR TITLE
Variable Experience Rollout Part 2 -- CPC|A

### PR DESCRIPTION
**Pull request by erikwijmans**
## Motivation and Context

Implementation of Action-Conditional Contrastive Predictive Coding (CPC|A) that works with the mini-batches generated by VER. This is a useful aux loss that helps with visual representation learning. I've been tempted to add an implementation for a while, so it made sense to add an implementation that works with VER.

This also includes a more general action embedding class.

## How Has This Been Tested

The test

## Types of changes

- New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: habitat_baselines/rl/models/action_embedding.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    def forward(self, action, masks=None):
+        action = action.long() + 1
+        if masks is not None:
+            action.masked_fill_(torch.logical_not(masks), 0)
+
+        return self.embedding(action.squeeze(-1))
+
+
+class ActionEmbedding(nn.Module):
```
Comment (ASzot) [inline habitat_baselines/rl/models/action_embedding.py:94]: Great, this is very helpful code in general.

File: habitat_baselines/rl/ppo/policy.py [hunk 2]
```diff
@@ -171,6 +208,13 @@ def all_policy_tensors(self):
         for c in self.policy_components:
             yield from c.buffers()
 
+    def aux_loss_parameters(self):
+        res = {}
+        for k, v in self.aux_loss_modules.items():
+            res[k] = v.parameters()
+
+        return res
```
Comment (Skylion007) [inline habitat_baselines/rl/ppo/policy.py:216]:
```suggestion
        return {k: v.parameters() for k, v in self.aux_loss_modules.items()}
```
faster and one liner.

**Review by Skylion007 — APPROVED** [review]
Looks good to me. One minor nit.

**Comment by erikwijmans** [comment]
@ASzot 
image

Blue is with CPC|A, orange is normal. This is SPL on PointNav with an RGB camera on HM3D. Pointnav is really stable in my experience so this is significant.

File: habitat_baselines/config/default.py [hunk 3]
```diff
@@ -174,6 +175,19 @@
 # policy inference time during rollout generation
 # Not that this does not change the memory requirements
 _C.RL.PPO.use_double_buffered_sampler = False
+# VER
```
Comment (mathfac) [inline habitat_baselines/config/default.py:178]:
```suggestion
# Variable Experience Rollout
```

File: habitat_baselines/config/default.py [hunk 4]
```diff
@@ -174,6 +175,19 @@
...
+_C.RL.VER = CN()
+_C.RL.VER.variable_experience = True
+_C.RL.VER.num_inference_workers = 2
+_C.RL.VER.overlap_rollouts_and_learn = False
+## AUX
+_C.RL.auxiliary_losses = CN()
+_C.RL.auxiliary_losses.enabled = []
+_C.RL.auxiliary_losses.cpca = CN()
```
Comment (mathfac) [inline habitat_baselines/config/default.py:186]:
```suggestion
#  Action-Conditional Contrastive Predictive Coding Loss
_C.RL.auxiliary_losses.cpca = CN()
```
